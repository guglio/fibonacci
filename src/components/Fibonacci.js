import React, { Component } from 'react';

import FibonacciNth from '../containers/FibonacciNth';

class Fibonacci extends Component {
  constructor(props) {
    super(props);
    this.state = {
      fibonacciN : 0,
      result : 0,
    }
  }

  handleChange(state){
    this.setState(state);
  }

  calculareFibonacci(e){
    e.preventDefault();
    let result = FibonacciNth(this.state.fibonacciN);
    this.handleChange({result});
  }

  render() {
    const { fibonacciN, result} = this.state;
    return (
      <div>
        <form onSubmit={(e) => this.calculareFibonacci(e)} id="formFibonacci">
          <input type='number' value={fibonacciN} onChange={(e) => this.handleChange({fibonacciN : e.target.value})} id='fibonacciNum'/>
          <button type='submit' id='submit'>Calculate</button>
        </form>
        <div id='FibonacciNth'>{result}</div>
      </div>
    )
  }
};

export default Fibonacci;
