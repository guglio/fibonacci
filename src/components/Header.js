import React from 'react';

const Header = () => (
  <header>
    <h1>Fibonacci Calculator</h1>
  </header>
);

export default Header;
