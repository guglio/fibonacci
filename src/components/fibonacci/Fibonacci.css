.fibonacci{
  padding: 0 calc(var(--baseline) * 2);
  margin-top: calc(var(--baseline) * 3);
}
.fibonacci input,
.fibonacci button{
  box-sizing: border-box;
  -moz-box-sizing: border-box;
}
.input{
  width: 280px;
  font-size: calc(var(--baseline) * 2);
  padding: calc(var(--baseline) * 1) calc(var(--baseline) * 2);
}
input[type=number]::-webkit-inner-spin-button,
input[type=number]::-webkit-outer-spin-button {
  -webkit-appearance: none;
  margin: 0;
}
.btn-group{
  padding: calc(var(--baseline) * 2) 0;
}
.btn-group > .btn{
  margin-right: calc(var(--baseline) * 2);
}
.btn{
  padding: calc(var(--baseline) * 1) calc(var(--baseline) * 2);
  border: 1px solid;
  font-size: calc(var(--baseline) * 2);
  transition: background-color .3s, color .3s, border-color .3s;
  cursor: pointer;
}
.btn-submit{
  border-color: var(--primary-dark);
  color: var(--primary-dark);
}
.btn-submit:hover{
  background-color: var(--primary-main);
  color:var(--primary-text);
  border-color: var(--primary-light);
}
.btn-reset{
  border-color:var(--secondary-dark);
  color:var(--secondary-dark);
}
.btn-reset:hover{
  background-color: var(--secondary-main);
  color:var(--secondary-text);
  border-color: var(--secondary-light);
}
