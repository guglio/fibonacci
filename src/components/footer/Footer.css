.footer{
  padding: 0 16px calc(16px * 4) 16px;
  position:absolute;
  bottom:0;
}
.credits > a{
  text-decoration: underline;
  color: inherit;
}
