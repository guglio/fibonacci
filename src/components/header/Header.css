.header{
  height: 56px;
  background-color: #3949ab;
  padding: 0 16px;
}
.title{
  margin: 0;
  color: #fff;
  line-height: 56px;
}
