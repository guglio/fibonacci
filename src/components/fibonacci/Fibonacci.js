import React, { Component } from 'react';

// function to calculate the Fibonacci sequence
import FibonacciNth from '../../containers/FibonacciNth';
// errors output
import { error } from '../../errors';

import './Fibonacci.css';

class Fibonacci extends Component {
  constructor(props) {
    super(props);
    this.state = {
      fibonacciN : 2, // integer entered from the user
      message : '', // output message for the user
    }
  }

  // handle the state change
  handleChange(state){
    this.setState(state);
  }

  // validate field in the form
  // if the input is not a valid one, return an error message
  validateForm(e){
    // avoid the default form submit event
    e.preventDefault();
    // shorter variable name and more readable
    let num = this.state.fibonacciN;
    // if the numer is less than 2, return false
    let numInRange = num >= 2;
    // create a custom error message
    let errorMessage = numInRange ? false : error.NUM_NOT_IN_RANGE;

    // if the errorMessage exists, don't calculate
    // but show the error message as result
    errorMessage ?
      this.handleChange({message : `Error: ${errorMessage}`})
      : this.calculareFibonacci(num);
  }

  // function called after the input is validated
  // that calculate the Fibonacci nth number
  calculareFibonacci(){
    // shorter variable name and more readable
    let num = this.state.fibonacciN;
    // calculate the Fibonacci nth number
    let result = FibonacciNth(num);
    // set the result's message for the user
    let message = `F(${num}) = ${result}`;
    // save the calculation into a state to display to the user
    this.handleChange({message});
  }

  // reset the form to the original state
  resetForm(){
    this.handleChange({fibonacciN : 2, message: ''});
  }

  render() {
    const { fibonacciN, message} = this.state;

    return (
      <div className="fibonacci">
        <form
          onSubmit={(e) => this.validateForm(e)}
          id="formFibonacci"
        >
          <input
            type='number'
            id='fibonacciNum'
            className='input input-number'
            value={fibonacciN}
            onChange={(e) => this.handleChange({fibonacciN : e.target.value})}
            autoComplete='off'
          />
          <div className="btn-group">
            <button
              type='submit'
              id='submit'
              className="btn btn-submit"
            >
              Calculate
            </button>
            <button
              type='reset'
              id='reset'
              className="btn btn-reset"
              onClick={() => this.resetForm()}
            >
              Reset
            </button>
          </div>
        </form>
        <div id='FibonacciNth' className='result'>
          <p>{message}</p>
        </div>
      </div>
    )
  }
};

export default Fibonacci;
