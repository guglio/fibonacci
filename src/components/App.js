import React, { Component } from 'react';

import Header from './header/Header';
import Fibonacci from './Fibonacci';
import Footer from './footer/Footer';

class App extends Component {
  render() {
    return (
      <div>
        <Header />
        <Fibonacci />
        <Footer />
      </div>
    );
  }
}

export default App;
