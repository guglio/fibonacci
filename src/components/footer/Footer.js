import React from 'react';

import './Footer.css';

const Footer = () => (
  <footer className="footer">
    <p className="credits">Created by <a href="//guglielmo-turco.com" target="_blank" rel="noopener noreferrer">Guglielmo Turco</a></p>
  </footer>
);

export default Footer;
