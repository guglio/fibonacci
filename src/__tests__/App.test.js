import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import App from '../components/App';

import Header from '../components/Header';
import Fibonacci from '../components/Fibonacci';
import Footer from '../components/Footer';

describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  describe('Render all child components', () => {
    it('Header', () => {
      const header = shallow(<Header />);
       expect(header).toMatchSnapshot();
    });
    it('Fibonacci', () => {
      const fibonacci = shallow(
         <Fibonacci />
       );
       expect(fibonacci).toMatchSnapshot();
    });
    it('Footer', () => {
      const footer = shallow(
         <Footer />
       );
       expect(footer).toMatchSnapshot();
    });
  });
});
