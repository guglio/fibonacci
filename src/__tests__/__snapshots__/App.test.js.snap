// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Render all child components Fibonacci 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Fibonacci />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <form
          id="formFibonacci"
          onSubmit={[Function]}
>
          <input
                    id="fibonacciNum"
                    onChange={[Function]}
                    type="number"
                    value={2}
          />
          <button
                    id="submit"
                    type="submit"
          >
                    Calculate
          </button>
</form>,
        <div
          id="FibonacciNth"
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              id="fibonacciNum"
              onChange={[Function]}
              type="number"
              value={2}
/>,
            <button
              id="submit"
              type="submit"
>
              Calculate
</button>,
          ],
          "id": "formFibonacci",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "fibonacciNum",
              "onChange": [Function],
              "type": "number",
              "value": 2,
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Calculate",
              "id": "submit",
              "type": "submit",
            },
            "ref": null,
            "rendered": "Calculate",
            "type": "button",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": 0,
          "id": "FibonacciNth",
        },
        "ref": null,
        "rendered": 0,
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <form
            id="formFibonacci"
            onSubmit={[Function]}
>
            <input
                        id="fibonacciNum"
                        onChange={[Function]}
                        type="number"
                        value={2}
            />
            <button
                        id="submit"
                        type="submit"
            >
                        Calculate
            </button>
</form>,
          <div
            id="FibonacciNth"
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                id="fibonacciNum"
                onChange={[Function]}
                type="number"
                value={2}
/>,
              <button
                id="submit"
                type="submit"
>
                Calculate
</button>,
            ],
            "id": "formFibonacci",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "fibonacciNum",
                "onChange": [Function],
                "type": "number",
                "value": 2,
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Calculate",
                "id": "submit",
                "type": "submit",
              },
              "ref": null,
              "rendered": "Calculate",
              "type": "button",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": 0,
            "id": "FibonacciNth",
          },
          "ref": null,
          "rendered": 0,
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`App Render all child components Footer 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Footer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "Footer test",
    },
    "ref": null,
    "rendered": "Footer test",
    "type": "footer",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Footer test",
      },
      "ref": null,
      "rendered": "Footer test",
      "type": "footer",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`App Render all child components Header 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Header />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <h1>
        Fibonacci Calculator
</h1>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Fibonacci Calculator",
      },
      "ref": null,
      "rendered": "Fibonacci Calculator",
      "type": "h1",
    },
    "type": "header",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <h1>
          Fibonacci Calculator
</h1>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Fibonacci Calculator",
        },
        "ref": null,
        "rendered": "Fibonacci Calculator",
        "type": "h1",
      },
      "type": "header",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`App renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Header />,
        <Fibonacci />,
        <Footer />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Header />,
          <Fibonacci />,
          <Footer />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
